docker-compose.yaml

```diff
@@ -0,0 +1,106 @@
+version: '3.8'
+services:
+  setup:
+    profiles: [ "setup" ]
+    image: keygen/api:latest
+    environment:
+      DATABASE_URL: "postgres://${POSTGRES_USER}:${POSTGRES_PASSWORD}@postgres:5432/${POSTGRES_DB}"
+      DISABLE_DATABASE_ENVIRONMENT_CHECK: 1
+    env_file:
+      - .env
+    command: >
+      sh -c "
+        until pg_isready -h postgres -U ${POSTGRES_USER}; do
+          echo 'Waiting for PostgreSQL...'
+          sleep 1
+        done
+        bundle exec rails keygen:setup
+      "
+    depends_on: [ postgres ]
+    tty: true
+    stdin_open: true
+
+  postgres:
+    image: postgres:9.6
+    ports:
+      - "5432:5432"
+    environment:
+      POSTGRES_USER: ${POSTGRES_USER}
+      POSTGRES_PASSWORD: ${POSTGRES_PASSWORD}
+      POSTGRES_DB: ${POSTGRES_DB}
+    volumes:
+      - postgres:/var/lib/postgresql/data
+
+  redis:
+    image: redis
+    ports:
+      - "6379:6379"
+    volumes:
+      - redis:/data
+
+  web:
+    image: keygen/api:latest
+    ports:
+      - "3000:3000"
+    networks:
+      - default
+      - web_network
+    command: "web"
+    env_file:
+      - .env
+    environment:
+      DATABASE_URL: "postgres://${POSTGRES_USER}:${POSTGRES_PASSWORD}@postgres:5432/${POSTGRES_DB}"
+    volumes:
+      - keygen:/etc/keygen
+    depends_on:
+      - postgres
+      - redis
+
+  worker:
+    image: keygen/api:latest
+    command: "worker"
+    env_file:
+      - .env
+    environment:
+      DATABASE_URL: "postgres://${POSTGRES_USER}:${POSTGRES_PASSWORD}@postgres:5432/${POSTGRES_DB}"
+    volumes:
+      - keygen:/etc/keygen
+    depends_on:
+      - postgres
+      - redis
+
+  caddy:
+    image: caddy:latest
+    restart: unless-stopped
+    ports:
+      - "80:80"
+      - "443:443"
+    volumes:
+      - caddy_data:/data
+      - caddy_config:/config
+    env_file:
+      - .env
+    command: >
+      sh -c '
+        echo "${KEYGEN_HOST} ${KEYGEN_HOSTS} {
```

Comment:
Will this work even though `KEYGEN_HOSTS` is a comma-delimited list?

Reply:
Fails for me:

```
caddy-1     | {"level":"info","ts":1726262644.073456,"msg":"using config from file","file":"/etc/caddy/Caddyfile"}                                                                                                                                                                                                         
caddy-1     | Error: adapting config using caddyfile: Site addresses cannot contain a comma ',': 'auth.keygen.localhost,get.keygen.localhost' - put a space after the comma to separate site addresses, at /etc/caddy/Caddyfile:1
caddy-1     | {"level":"info","ts":1726262647.809099,"msg":"using config from file","file":"/etc/caddy/Caddyfile"}                                                                                                                                                                                                         
caddy-1     | Error: adapting config using caddyfile: Site addresses cannot contain a comma ',': 'auth.keygen.localhost,get.keygen.localhost' - put a space after the comma to separate site addresses, at /etc/caddy/Caddyfile:1
caddy-1     | {"level":"info","ts":1726262654.78514,"msg":"using config from file","file":"/etc/caddy/Caddyfile"}                                                                                                                                                                                                          
caddy-1     | Error: adapting config using caddyfile: Site addresses cannot contain a comma ',': 'auth.keygen.localhost,get.keygen.localhost' - put a space after the comma to separate site addresses, at /etc/caddy/Caddyfile:1
```

Reply:
You can use variable expansion, `${KEYGEN_HOSTS/,/ }`.

Reply:
Just thinking out loud here — would it be cleaner to just add an actual `Caddyfile` to the repo? That way people can easily customize it if needed. I think Caddy supports env vars in the Caddyfile.

Reply:
KEYGEN_HOSTS is a blocker tho, it throws the same error (put a space after the comma to separate site addresses).

I got with working with shell script

Reply:
```    command: >
      sh -c '
        echo "${KEYGEN_HOST}$${KEYGEN_HOSTS:+,} $${KEYGEN_HOSTS//,/, } {
          tls internal
          reverse_proxy web:3000
        }" > /etc/caddy/Caddyfile &&
        caddy run --config /etc/caddy/Caddyfile
      '```